File: python_api/SparkVersion/ml/ClassificatorsAndRegressors.py
from SparkVersion.ml.regression.Regressors import Regressors
from SparkVersion.ml.classification.Classificators import Classificators


class ClassificatorsAndRegressors:
    def __init__(self, data, train, test, maxCategories=4):
        self.data = data
        self.train = train
        self.test = test
        self.maxCategories = maxCategories

    def chooseTypeOfProblem(self, typeOfProblem):
        if typeOfProblem == "Regression":
            return Regressors(self.data, self.train, self.test)
        else:
            return Classificators(self.data, self.train, self.test)

File: scala_api/src/main/scala/application/dataAnalysis/ml/ClassificatorsAndRegressors.scala
package application.dataAnalysis.ml

import application.dataAnalysis.Context._
import application.dataAnalysis.interfaces.ModelCreateProcess
import application.dataAnalysis.ml.classification.Classificators
import application.dataAnalysis.ml.regression.Regressors
import org.apache.spark.ml.feature.{VectorIndexer, VectorIndexerModel}
import org.apache.spark.ml.{Pipeline, PipelineModel}
import org.apache.spark.sql.{DataFrame, Dataset, Row}


class ClassificatorsAndRegressors(
  val data: DataFrame,
  val train: Dataset[Row],
  val test: Dataset[Row],
  val maxCategories: Option[Int] = None
) {

  val pipeline = new Pipeline()

  def chooseTypeOfProblem(typeOfProblem: String): ClassificatorsAndRegressors with ModelCreateProcess = {
    typeOfProblem match {
      case "Regression" => Regressors(data, train, test)
      case _ => Classificators(data, train, test)
    }
  }

  val featureIndexer: VectorIndexerModel = {
    new VectorIndexer()
      .setInputCol("features")
      .setOutputCol("indexedFeatures")
      .setMaxCategories(maxCategories.getOrElse(4))
      .fit(data)
  }

  def safeResources(pipelineModel: PipelineModel): Unit = {
    pipelineModel.write.overwrite().save(getModelFilePath)
  }
}